pkg/pool-utils/contracts/rates/PriceRateCache.sol

```diff
@@ -23,55 +23,71 @@ import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
  * useful for slow changing rates, such as those that arise from interest-bearing tokens (e.g. waDAI into DAI).
  *
  * The cache data is packed into a single bytes32 value with the following structure:
- * [   expires   | duration | price rate value ]
- * [   uint64    |  uint64  |      uint128     ]
- * [ MSB                                   LSB ]
+ * [ 32 bits |  32 bits  |   96 bits  |    96 bits    ]
+ * [ expires | duration  | last rate  | current rate  ]
+ *  |MSB                                           LSB|
  *
- *
- * 'rate' is an 18 decimal fixed point number, supporting rates of up to ~3e20. 'expires' is a Unix timestamp, and
+ * 'rate' is an 18 decimal fixed point number, supporting rates of up to ~3e10. 'expires' is a Unix timestamp, and
  * 'duration' is expressed in seconds.
  */
 library PriceRateCache {
     using WordCodec for bytes32;
 
-    uint256 private constant _PRICE_RATE_CACHE_VALUE_OFFSET = 0;
-    uint256 private constant _PRICE_RATE_CACHE_DURATION_OFFSET = 128;
-    uint256 private constant _PRICE_RATE_CACHE_EXPIRES_OFFSET = 128 + 64;
+    uint256 private constant _CURRENT_PRICE_RATE_OFFSET = 0;
+    uint256 private constant _POST_JOINEXIT_RATE_OFFSET = 96;
+    uint256 private constant _PRICE_RATE_CACHE_DURATION_OFFSET = 192;
+    uint256 private constant _PRICE_RATE_CACHE_EXPIRES_OFFSET = 224;
 
     /**
-     * @dev Returns the rate of a price rate cache.
+     * @dev Returns the current rate of a price rate cache.
      */
     function getRate(bytes32 cache) internal pure returns (uint256) {
-        return cache.decodeUint(_PRICE_RATE_CACHE_VALUE_OFFSET, 128);
+        return cache.decodeUint(_CURRENT_PRICE_RATE_OFFSET, 96);
+    }
+
+    /**
+     * @dev Returns the current rate of a price rate cache.
+     */
+    function getPostJoinExitRate(bytes32 cache) internal pure returns (uint256) {
+        return cache.decodeUint(_POST_JOINEXIT_RATE_OFFSET, 96);
+    }
+
+    /**
+     * @dev Sets the post join/exit rate in an existing cache record, and returns it.
+     */
+    function setPostJoinExitRate(bytes32 cache, uint256 rate) internal pure returns (bytes32) {
+        _require(rate >> 96 == 0, Errors.PRICE_RATE_OVERFLOW);
+
+        return cache.insertUint(rate, _POST_JOINEXIT_RATE_OFFSET, 96);
     }
 
     /**
      * @dev Returns the duration of a price rate cache.
      */
     function getDuration(bytes32 cache) internal pure returns (uint256) {
-        return cache.decodeUint(_PRICE_RATE_CACHE_DURATION_OFFSET, 64);
+        return cache.decodeUint(_PRICE_RATE_CACHE_DURATION_OFFSET, 32);
```

Comment:
Let's make these numbers be constants (e.g. DURATION_SIZE), and then build the offsets out of them (RATE_OFFSET = 0, OLD_RATE_OFFSET = RATE_OFFSET + RATE_DURATION, etc).

Reply:
I saw the old code was like this, but we don't do this anywhere else we have these offset constants (e.g., ManagedPools). To me doing it that way seems brittle. If you change the order, you have to go through and fix all the arithmetic everywhere, vs just changing the offsets in one place and everything works.

I could see making the bitLengths constants.

Reply:
It'd be a change in a single place though, wouldn't it? Each field simply specifies the preceding one. But yeah, let's at least make the lenghts constant.

Reply:
Ok; already did that.